scipy/stats/_qmc.py

```diff
@@ -898,6 +988,151 @@ def random(self, n: IntNumber = 1) -> np.ndarray:
+
+    >>> from scipy.stats import qmc
+    >>> sampler = qmc.OptimalLatinHypercube(d=2)
+    >>> sample = sampler.random(n=5)
+    >>> sample
+    array([[0.84052691, 0.53664833],  # random
+           [0.1545328 , 0.36265316],
+           [0.52177809, 0.93343721],
+           [0.68033825, 0.06474907],
+           [0.26544879, 0.61163943]])
+
+    Compute the quality of the sample using the discrepancy criterion.
+
+    >>> qmc.discrepancy(sample)
+    0.01602810655510689  # random
+
+    You can possibly improve the quality of the sample by performing more
+    optimization iterations by using `niter`:
+
+    >>> sampler_2 = qmc.OptimalLatinHypercube(d=2, n_perturbations=5000)
+    >>> sample_2 = sampler_2.random(n=5)
+    >>> qmc.discrepancy(sample_2)
+    0.01600398742881648  # random
+
+    """
+
+    def __init__(self, d: IntNumber, *, start_sample: npt.ArrayLike = None,
+                 n_perturbations: IntNumber = 10000,
+                 seed: SeedType = None):
+        super().__init__(d=d, seed=seed)
+        self.n_perturbations = n_perturbations
+        self.start_sample: Optional[np.ndarray] = None
+
+        if start_sample is not None:
+            self.start_sample = np.asarray(start_sample).copy()
+            self.start_disc = discrepancy(self.start_sample)
+        else:
+            self.start_disc = np.inf
+
+        self.lhs = LatinHypercube(self.d, seed=self.rng)
+
+    def random(self, n: IntNumber = 1):
+        """Draw `n` in the half-open interval ``[0, 1)``.
+
+        Parameters
+        ----------
+        n : int, optional
+            Number of samples to generate in the parameter space. Default is 1.
+
+        Returns
+        -------
+        sample : array_like (n, d)
+            Optimal LHS sample.
+
+        """
+        if self.d == 0 or n == 0:
+            return np.empty((n, self.d))
+
+        if self.start_sample is None:
+            best_sample = self.lhs.random(n)
+            best_disc = discrepancy(best_sample)
+        else:
+            best_sample = self.start_sample
+            best_disc = self.start_disc
+
+        if n == 1:
+            return best_sample
+
+        bounds = ([0, self.d - 1],
+                  [0, n - 1],
+                  [0, n - 1])
+
+        for _ in range(self.n_perturbations):
+            col = rng_integers(self.rng, *bounds[0])
+            row_1 = rng_integers(self.rng, *bounds[1])
+            row_2 = rng_integers(self.rng, *bounds[2])
+            doe = best_sample.copy()
+            disc = _perturb_discrepancy(best_sample,
+                                        row_1, row_2, col,
+                                        best_disc)
+            if disc < best_disc:
+                doe[row_1, col], doe[row_2, col] = (
+                    doe[row_2, col], doe[row_1, col])
+                best_disc = disc
+                best_sample = doe
+
+        self.num_generated += n
+        return best_sample
+
+    def reset(self):
```

Comment:
`num_generated` is private, and it's not part of the algorithm, so if no `start_sample` is provided, `reset` has no meaningful effect, right? 

Er, `num_generated` is essentially undocumented... it probably _should_ be private and marked as such with a preceding underscore. 

I see it mentioned in the [`QMCEngine`](https://33172-1460385-gh.circle-artifacts.com/0/html-scipyorg/reference/generated/scipy.stats.qmc.QMCEngine.html#scipy.stats.qmc.QMCEngine) documentation. This reminds me of gh-13225. It can be a headache when there are pseudo-public parameters that users start relying on.

Reply:
> `num_generated` is private, and it's not part of the algorithm, so if no `start_sample` is provided, `reset` has no meaningful effect, right?

True! I can remove this function and put the `self.lhs = LatinHypercube(self.d, seed=self.rng)` directly in `self.random`.

> Er, `num_generated` is essentially undocumented... it probably _should_ be private and marked as such with a preceding underscore.
> 
> I see it mentioned in the [`QMCEngine`](https://33172-1460385-gh.circle-artifacts.com/0/html-scipyorg/reference/generated/scipy.stats.qmc.QMCEngine.html#scipy.stats.qmc.QMCEngine) documentation. This reminds me of [gh-13225](https://github.com/scipy/scipy/issues/13225). It can be a headache when there are pseudo-public parameters that users start relying on.

Yes it is in the doc. I kind off wanted to have it as most of these methods are iterative and it makes sense to be able to access the state.

LHS is particular in a sense that it is a one shot algorithm. You give n and you are done with it. But there are methods in the literature to extend make LHS iterative. I never tried to code anything here though and did not find any public code. So I don't know how it performs.

Shall I add a note in the doc to warn about this attribute is public and we should be careful not to remove it? (or you prefer the other way around? private and we warn (or not) not to use it?)

Reply:
My opinion would be to either go all in and document it as public or change it to `_num_generated` for 1.7.1 before there has been much time for people to notice and start relying on it. I suppose I would prefer making it private until we can think of good reasons for exposing it.

Reply:
I propose to do this in a follow up and change this in all classes if still wanted.